README.md
# Triglav::Agent::Vertica

Triglav Agent for Vertica

## Requirements

* Ruby >= 2.3.0

## Prerequisites

* Vertica table must have a DATE column for `daily` resource monitor
* Vertica table must have a TIMESTAMP or TIMESTAMPTZ column for `hourly` resource monitor
* Vertica view is not supported (since `epoch` column can not be retrieved)

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'triglav-agent-vertica'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install triglav-agent-vertica

## CLI

```
Usage: triglav-agent-vertica [options]
    -c, --config VALUE               Config file (default: config.yml)
    -s, --status VALUE               Status stroage file (default: status.yml)
    -t, --token VALUE                Triglav access token storage file (default: token.yml)
        --dotenv                     Load environment variables from .env file (default: false)
    -h, --help                       help
        --log VALUE                  Log path (default: STDOUT)
        --log-level VALUE            Log level (default: info)
```

Run as:

```
TRIGLAV_ENV=development bundle exec triglav-agent-vertica --dotenv -c config.yml
```

## Configuration

Prepare config.yml as [example/config.yml](./example/config.yml).

You can use erb template. You may load environment variables from .env file with `--dotenv` option as an [example/example.env](./example/example.env) file shows.

### serverengine section

You can specify any [serverengine](https://github.com/fluent/serverengine) options at this section

### triglav section

Specify triglav api url, and a credential to authenticate.

The access token obtained is stored into a token storage file (--token option).

### vertica section

This section is the special section for triglav-agent-vertica.

* **monitor_interval**: The interval to watch tables (number, default: 60)
* **connection_info**: key-value pairs of vertica connection info where keys are resource URI pattern in regular expression, and values are connection infomation

## How it behaves

1. Authenticate with triglav
  * Store the access token into the token storage file
  * Read the token from the token storage file next time
  * Refresh the access token if it is expired
2. Repeat followings in `monitor_interval` seconds:
3. Obtain resource (table) lists of the specified prefix (keys of connection_info) from triglav.
4. Connect to vertica with an appropriate connection info for a resource uri, and find tables which are newer than last check.
5. Store checking information into the status storage file for the next time check.

## Development

### Prepare

```
./prepare.sh
```

Edit `.env` or `config.yml` file directly.

### Start

Start up triglav api on localhost.

Run triglav-anget-vertica as:

```
TRIGLAV_ENV=development bundle exec triglav-agent-vertica --dotenv --debug -c example/config.yml
```

The debug mode with --debug option ignores the `last_epoch` value in status file.

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/triglav-workflow/triglav-agent-vertica. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.


## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).



CHANGELOG.md
# 1.0.0 (2017-04-14)

initial release


test/helper.rb
# frozen_string_literal: true

require 'test/unit'
require 'test/unit/rr'
require 'pry'
require 'timecop'
require 'triglav/agent/vertica'
require_relative 'support/create_table'

TEST_ROOT = __dir__
ROOT = File.dirname(__dir__)
ENV['APP_ENV'] = 'test'

opts = {
  config: File.join(TEST_ROOT, 'support', 'config.yml'),
  status: File.join(TEST_ROOT, 'tmp', 'status.yml'),
  token: File.join(TEST_ROOT, 'tmp', 'token.yml'),
  dotenv: true,
  debug: true,
}
$setting = Triglav::Agent::Configuration.setting_class.new(opts)
$logger = $setting.logger


test/test_worker.rb
# frozen_string_literal: true

require 'triglav/agent/vertica/worker'
require_relative 'helper'
require_relative 'support/stub_processor'

class TestWorker < Test::Unit::TestCase
  include StubProcessor

  class Worker
    include Triglav::Agent::Vertica::Worker
    define_method(:worker_id) { 1 }
  end

  def setup
    stub_processor
  end

  def worker
    Worker.new
  end

  def test_process_with_success
    assert_nothing_raised { worker.process }
  end
end


test/test_monitor.rb
# frozen_string_literal: true

require 'triglav/agent/vertica/monitor'
require_relative 'helper'
require_relative 'support/create_table'

# This test requires a real connection to vertica, now
# Configure .env to set proper connection_info of test/config.yml
#
# TRIGLAV_URL=http://localhost:7800
# TRIGLAV_USERNAME=triglav_test
# TRIGLAV_PASSWORD=triglav_test
# VERTICA_HOST=vertica
# VERTICA_PORT=5433
# VERTICA_DATABASE=vdb
# VERTICA_USER=dbwrite
# VERTICA_PASSWORD=xxxxxxx
if File.exist?(File.join(ROOT, '.env'))
  class TestMonitor < Test::Unit::TestCase
    include CreateTable
    Monitor = Triglav::Agent::Vertica::Monitor

    class << self
      def startup
        Timecop.travel(Time.parse("2016-12-30 23:00:00 +09:00"))
        create_table
        insert_data
      end

      def shutdown
        drop_table
        Timecop.return
      end
    end

    def build_resource(params = {})
      TriglavClient::ResourceResponse.new({
        uri: "vertica://#{host}:#{port}/#{db}/#{schema}/#{table}",
        unit: 'daily',
        timezone: '+09:00',
        span_in_days: 2,
        consumable: true,
        notifiable: false,
      }.merge(params))
    end

    def resource_uri_prefix
      'vertica://'
    end

    def test_process
      resource = build_resource
      monitor = Monitor.new(connection, resource_uri_prefix, resource)
      assert_nothing_raised { monitor.process }
    end

    def test_get_hourly_events
      resource = build_resource(unit: 'hourly')
      monitor = Monitor.new(connection, resource_uri_prefix, resource)
      success = monitor.process do |events|
        assert { events != nil}
        assert { events.size == resource.span_in_days * 24 }
        event = events.first
        assert { event.keys == %i[uuid resource_uri resource_unit resource_time resource_timezone payload] }
        assert { event[:resource_uri] == resource.uri }
        assert { event[:resource_unit] == resource.unit }
        assert { event[:resource_timezone] == resource.timezone }
      end
      assert { success }
    end

    def test_get_daily_events
      resource = build_resource(unit: 'daily')
      monitor = Monitor.new(connection, resource_uri_prefix, resource)
      success = monitor.process do |events|
        assert { events != nil}
        assert { events.size == resource.span_in_days }
        event = events.first
        assert { event.keys == %i[uuid resource_uri resource_unit resource_time resource_timezone payload] }
        assert { event[:resource_uri] == resource.uri }
        assert { event[:resource_unit] == resource.unit }
        assert { event[:resource_timezone] == resource.timezone }
      end
      assert { success }
    end

    def test_get_singular_events
      resource = build_resource(unit: 'singular')
      monitor = Monitor.new(connection, resource_uri_prefix, resource)
      success = monitor.process do |events|
        assert { events != nil}
        assert { events.size == 1 }
        event = events.first
        assert { event.keys == %i[uuid resource_uri resource_unit resource_time resource_timezone payload] }
        assert { event[:resource_uri] == resource.uri }
        assert { event[:resource_unit] == resource.unit }
        assert { event[:resource_timezone] == resource.timezone }
      end
      assert { success }
    end

    def test_get_daily_hourly_events
      resource = build_resource(unit: 'daily,hourly')
      monitor = Monitor.new(connection, resource_uri_prefix, resource)
      success = monitor.process do |events|
        assert { events != nil}
        assert { events.size == resource.span_in_days * 24 + resource.span_in_days }
        assert { events.first[:resource_unit] == 'hourly' }
        assert { events.last[:resource_unit] == 'daily' }
      end
      assert { success }
    end

    def test_get_singular_daily_hourly_events
      resource = build_resource(unit: 'singular,daily,hourly')
      monitor = Monitor.new(connection, resource_uri_prefix, resource)
      success = monitor.process do |events|
        assert { events != nil}
        assert { events.size == resource.span_in_days * 24 + resource.span_in_days + 1 }
        assert { events.any? {|e| e[:resource_unit] == 'hourly' } }
        assert { events.any? {|e| e[:resource_unit] == 'daily' } }
        assert { events.any? {|e| e[:resource_unit] == 'singular' } }
      end
      assert { success }
    end

    def test_query_date
      resource = build_resource(
        uri: "vertica://#{host}:#{port}/#{db}/#{schema}/#{table}?date=date"
      )
      monitor = Monitor.new(connection, resource_uri_prefix, resource)
      q_date = monitor.send(:q_date)
      assert { q_date == %Q["date"] }
    end

    def test_query_timestamp
      resource = build_resource(
        uri: "vertica://#{host}:#{port}/#{db}/#{schema}/#{table}?timestamp=timestamp"
      )
      monitor = Monitor.new(connection, resource_uri_prefix, resource)
      q_timestamp = monitor.send(:q_timestamp)
      assert { q_timestamp == %Q["timestamp"] }
    end

    # Value specification:
    # * A value looks like an integer string is treated as an integer.
    # * If you want to treat it as as string, surround with double quote or single quote.
    # * A value does not look like an integer is treated as a string.
    # Operator specification:
    # * Only equality is supported now
    def test_query_where
      resource = build_resource(
        uri: "vertica://#{host}:#{port}/#{db}/#{schema}/#{table}?where[id]=0&where[uuid]='0'&where[d]=2016-12-30"
      )
      monitor = Monitor.new(connection, resource_uri_prefix, resource)
      q_where = monitor.send(:q_where)
      assert { q_where == %Q["id" = 0 AND "uuid" = '0' AND "d" = '2016-12-30'] }
    end

    def test_get_events_with_query_where
      resource = build_resource(
        unit: 'singular,daily,hourly',
        uri: "vertica://#{host}:#{port}/#{db}/#{schema}/#{table}?where[id]=0&where[uuid]='0'"
      )
      monitor = Monitor.new(connection, resource_uri_prefix, resource)
      success = monitor.process do |events|
        assert { events != nil}
        assert { events.size == 3 }
        assert { events.any? {|e| e[:resource_unit] == 'hourly' } }
        assert { events.any? {|e| e[:resource_unit] == 'daily' } }
        assert { events.any? {|e| e[:resource_unit] == 'singular' } }
      end
      assert { success }
    end
  end
end


lib/triglav/agent/vertica/worker.rb
require 'triglav/agent/vertica/processor'

module Triglav::Agent
  module Vertica
    module Worker
      # serverengine interface
      def initialize
        @timer = Timer.new
      end

      # serverengine interface
      def reload
        $logger.info { "Worker#reload worker_id:#{worker_id}" }
        $setting.reload
      end

      # serverengine interface
      def run
        $logger.info { "Worker#run worker_id:#{worker_id}" }
        start
        until @stop
          @timer.wait(monitor_interval) { process }
        end
      rescue => e
        # ServerEngine.dump_uncaught_error does not tell me e.class
        log_error(e)
        raise e
      end

      def process
        started = Time.now
        $logger.info { "Start Worker#process worker_id:#{worker_id}" }

        total_count = 0
        total_success_count = 0
        resource_uri_prefixes.each do |resource_uri_prefix|
          break if stopped?
          processor = Processor.new(self, resource_uri_prefix)
          total_count += processor.total_count
          total_success_count += processor.process
        end

        elapsed = Time.now - started
        $logger.info {
          "Finish Worker#process worker_id:#{worker_id} " \
          "success_count/total_count:#{total_success_count}/#{total_count} elapsed:#{elapsed.to_f}sec"
        }
      end

      def start
        @timer.start
        @stop = false
      end

      # serverengine interface
      def stop
        $logger.info { "Worker#stop worker_id:#{worker_id}" }
        @stop = true
        @timer.stop
      end

      def stopped?
        @stop
      end

      private

      def log_error(e)
        $logger.error { "#{e.class} #{e.message} #{e.backtrace.join("\\n")}" } # one line
      end

      def monitor_interval
        $setting.dig(:vertica, :monitor_interval) || 60
      end

      def resource_uri_prefixes
        $setting.dig(:vertica, :connection_info).keys
      end
    end
  end
end


lib/triglav/agent/vertica/watcher.rb
require 'triglav/agent/vertica/connection'
require 'vertica'
require 'uri'

module Triglav::Agent::Vertica
  class Watcher
    attr_reader :connection

    def initialize(connection)
      @connection = connection
      # IMPORTANT ASSUMPTION: other processes does not modify status file
      @status = Triglav::Agent::StorageFile.load($setting.status_file)
    end

    def process(resource)
      last_epoch = get_last_epoch(resource)
      $logger.debug { "Start process #{resource.uri}, last_epoch:#{last_epoch}" }
      events, new_last_epoch = get_events(resource, last_epoch: $setting.debug? ? 0 : last_epoch)
      $logger.debug { "Finish process #{resource.uri}, last_epoch:#{last_epoch}, new_last_epoch:#{new_last_epoch}" }
      return if events.nil? || events.empty?
      yield(events) # send_message
      update_status_file(resource, new_last_epoch)
      [events, new_last_epoch]
    end

    # @param [TriglavClient::ResourceResponse] resource
    # resource:
    #   uri: vertica://host/database/schema/table
    #   unit: 'daily' or 'hourly', or 'daily,hourly'
    #   timezone: '+09:00'
    #   span_in_days: 32
    # @param [Ineger] last_epoch
    def get_events(resource, last_epoch:)
      if !%w[daily hourly daily,hourly].include?(resource.unit) ||
          resource.timezone.nil? || resource.span_in_days.nil?
        $logger.warn { "Broken resource: #{resource.to_s}" }
        return nil
      end

      now = Time.now.localtime(resource.timezone)
      dates = resource.span_in_days.times.map do |i|
        (now - (i * 86000)).strftime('%Y-%m-%d')
      end

      last_epoch ||= get_last_epoch(resource.uri)
      q_last_epoch = Vertica.quote(last_epoch)

      _, schema, table = URI.parse(resource.uri).path[1..-1].split('/')
      q_schema = Vertica.quote_identifier(schema)
      q_table  = Vertica.quote_identifier(table)

      q_date = Vertica.quote_identifier(date_column)
      q_timestamp = Vertica.quote_identifier(timestamp_column)

      # 'daily,hourly': qeury in hourly way, then merge events into daily in ruby
      unit = resource.unit == 'daily,hourly' ? 'hourly' : resource.unit

      begin
        case unit
        when 'hourly'
          sql = "select " \
            "#{q_date} AS d, DATE_PART('hour', #{q_timestamp}) AS h, max(epoch) " \
            "from #{q_schema}.#{q_table} " \
            "where #{q_date} IN ('#{dates.join("','")}') " \
            "group by d, h having max(epoch) > #{q_last_epoch} " \
            "order by d, h"
        when 'daily'
          sql = "select " \
            "#{q_date} AS d, 0 AS h, max(epoch) " \
            "from #{q_schema}.#{q_table} " \
            "where #{q_date} IN ('#{dates.join("','")}') " \
            "group by d having max(epoch) > #{q_last_epoch} " \
            "order by d"
        end

        $logger.debug { "Query: #{sql}" }
        result = connection.query(sql)

        events = build_events(result, resource, unit)
        if resource.unit == 'daily,hourly'
          daily_events = build_daily_events_from_hourly(result, resource)
          events.concat(daily_events)
        end

        new_last_epoch = latest_epoch || last_epoch
        [events, new_last_epoch]
      rescue Vertica::Error::QueryError => e
        $logger.warn { "#{e.class} #{e.message}" }
        nil
      end
    end

    private

    def latest_epoch(result)
      result.map {|row| row[2] }.max
    end

    def build_events(result, resource, unit = resource.unit)
      result.map do |row|
        date, hour, epoch = row[0], row[1], row[2]
        {
          resource_uri: resource.uri,
          resource_unit: unit,
          resource_time: date_hour_to_i(date, hour, resource.timezone),
          resource_timezone: resource.timezone,
          payload: {epoch: epoch},
        }
      end
    end

    def build_daily_events_from_hourly(result, resource)
      max_epoch_of = {}
      result.each do |row|
        date, hour, epoch = row[0], row[1], row[2]
        max_epoch_of[date] = [epoch, max_epoch_of[date] || 0].max
      end
      daily_events = max_epoch_of.map do |date, epoch|
        {
          resource_uri: resource.uri,
          resource_unit: 'daily',
          resource_time: date_hour_to_i(date, 0, resource.timezone),
          resource_timezone: resource.timezone,
          payload: {epoch: epoch},
        }
      end
    end

    def date_hour_to_i(date, hour, timezone)
      Time.strptime("#{date.to_s} #{hour.to_i} #{timezone}", '%Y-%m-%d %H %z').to_i
    end

    def update_status_file(resource, last_epoch)
      Triglav::Agent::StorageFile.open($setting.status_file) do |fp|
        @status = fp.load # reload during locking
        update_status(resource, last_epoch)
        fp.dump(@status)
      end
    end

    def update_status(resource, last_epoch)
      (@status[:last_epoch] ||= {})[resource.uri.to_sym] = last_epoch
    end

    def get_last_epoch(resource)
      unless last_epoch = @status.dig(:last_epoch, resource.uri.to_sym)
        last_epoch = get_current_epoch
        update_status_file(resource, last_epoch)
      end
      last_epoch
    end

    def get_current_epoch
      connection.query('select GET_CURRENT_EPOCH()').first.first
    end

    def date_column
      $setting.dig(:vertica, :date_column) || 'd'
    end

    def timestamp_column
      $setting.dig(:vertica, :timestamp_column) || 't'
    end
  end
end


lib/triglav/agent/vertica/processor.rb
require 'triglav/agent/vertica/monitor'
require 'triglav/agent/vertica/connection'
require 'triglav/agent/vertica/error'
require 'parallel'
require 'connection_pool'

module Triglav::Agent
  module Vertica
    class Processor
      attr_reader :worker, :resource_uri_prefix

      def initialize(worker, resource_uri_prefix)
        @worker = worker
        @resource_uri_prefix = resource_uri_prefix
        @connection_pool = ConnectionPool.new(connection_pool_opts) {
          Connection.new(get_connection_info(resource_uri_prefix))
        }
        @api_client_pool = ConnectionPool.new(connection_pool_opts) {
          ApiClient.new # renew connection
        }
        @mutex = Mutex.new
      end

      def self.max_consecuitive_error_count
        3
      end

      def process
        success_count = 0
        consecutive_error_count = 0
        Parallel.each(resources, parallel_opts) do |resource|
          raise Parallel::Break if stopped?
          events = nil
          begin
            @connection_pool.with do |connection|
              monitor = Monitor.new(connection, resource, last_epoch: $setting.debug? ? 0 : nil)
              monitor.process do |_events|
                events = _events
                $logger.info { "send_messages:#{events.map(&:to_hash).to_json}" }
                @api_client_pool.with {|api_client| api_client.send_messages(events) }
              end
            end
            @mutex.synchronize do
              success_count += 1
              consecutive_error_count = 0
            end
          rescue => e
            log_error(e)
            $logger.info { "failed_events:#{events.map(&:to_hash).to_json}" } if events
            @mutex.synchronize do
              raise TooManyError if (consecutive_error_count += 1) > self.class.max_consecuitive_error_count
            end
          end
        end
        success_count
      end

      def total_count
        resources.size
      end

      private

      def resources
        return @resources if @resources
        @resources = ApiClient.new.list_aggregated_resources(resource_uri_prefix) || []
        $logger.debug { "resource_uri_prefix:#{resource_uri_prefix} resources.size:#{@resources.size}" }
        @resources
      end

      def parallel_size
        $setting.dig(:vertica, :parallel, :size) || 1
      end

      def parallel_type
        $setting.dig(:vertica, :parallel, :type) || 'thread'
      end

      def parallel_opts
        parallel_type == 'process' ? {in_processes: parallel_size} : {in_threads: parallel_size}
      end

      def connection_pool_size
        $setting.dig(:vertica, :connection_pool, :size) || parallel_size
      end

      def connection_pool_timeout
        $setting.dig(:vertica, :connection_pool, :timeout) || 60
      end

      def connection_pool_opts
        {size: connection_pool_size, timeout: connection_pool_timeout}
      end

      def get_connection_info(resource_uri_prefix)
        $setting.dig(:vertica, :connection_info)[resource_uri_prefix]
      end

      def log_error(e)
        $logger.error { "#{e.class} #{e.message} #{e.backtrace.join("\\n")}" } # one line
      end

      def stopped?
        worker.stopped? if worker
      end
    end
  end
end


lib/triglav/agent/vertica/connection.rb
require 'triglav/agent/base/connection'
require 'vertica'
require 'uri'

module Triglav::Agent
  module Vertica
    class Connection < Base::Connection
      attr_reader :connection_info

      def initialize(connection_info)
        @connection_info = connection_info
      end

      def close
        @connection.close rescue nil if @connection
      end

      def query(sql)
        connection.query(sql)
      end

      private

      def connection
        return @connection if @connection
        begin
          @connection = ::Vertica.connect(connection_params)
        rescue => e
          $logger.error { "Failed to connect #{connection_info[:host]}:#{connection_info[:port]} with #{connection_info[:user]}" }
          raise e
        end
        $logger.info { "Connected to #{connection_info[:host]}:#{connection_info[:port]}" }
        set_resource_pool
        set_memorycap
        @connection
      end

      def set_resource_pool
        if @connection_info[:resource_pool] and !@connection_info[:resource_pool].empty?
          @connection.query("set session resource_pool = '#{@connection_info[:resource_pool]}'")
        end
      end

      def set_memorycap
        if @connection_info[:memorycap] and !@connection_info[:memorycap].empty?
          @connection.query("set session memorycap = '#{@connection_info[:memorycap]}'")
        end
      end

      def connection_params
        params = @connection_info.dup
        params.delete(:resource_pool)
        params.delete(:memorycap)
        params.merge!(global_connection_params)
      end

      def global_connection_params
        params = {}
        params[:interruptable] = $setting.dig(:vertica, :interruptable) if $setting[:vertica].has_key?(:interruptable)
        params[:read_timeout] = $setting.dig(:vertica, :read_timeout) if $setting[:vertica].has_key?(:read_timeout)
        params
      end
    end
  end
end


lib/triglav/agent/vertica/monitor.rb
require 'triglav/agent/base/monitor'
require 'vertica'
require 'uri'
require 'cgi'
require 'securerandom'
require 'rack/utils'

module Triglav::Agent
  module Vertica
    class Monitor < Base::Monitor
      attr_reader :connection, :resource_uri_prefix, :resource, :periodic_last_epoch, :singular_last_epoch

      # @param [Triglav::Agent::Vertica::Connection] connection
      # @param [String] resource_uri_prefix
      # @param [TriglavClient::ResourceResponse] resource
      # resource:
      #   uri: vertica://host/database/schema/table
      #   unit: 'daily', 'hourly', 'singular', or their combinations such as 'singular,daily,hourly'
      #   timezone: '+09:00'
      #   span_in_days: 32
      #
      # View is not supported
      def initialize(connection, resource_uri_prefix, resource)
        @connection = connection
        @resource_uri_prefix = resource_uri_prefix
        @resource = resource
        @status = Triglav::Agent::Status.new(resource_uri_prefix, resource.uri)
        @periodic_last_epoch = $setting.debug? ? 0 : get_from_status_file(:periodic_last_epoch)
        @singular_last_epoch = $setting.debug? ? 0 : get_from_status_file(:singular_last_epoch)
      end

      def process
        unless resource_valid?
          $logger.warn { "Broken resource: #{resource.to_s}" }
          return nil
        end

        $logger.debug {
          msgs = ["Start process #{resource.uri}"]
          msgs << "periodic_last_epoch:#{periodic_last_epoch}" if periodic_last_epoch
          msgs << "singular_last_epoch:#{singular_last_epoch}" if singular_last_epoch
          msgs.join(', ')
        }

        if periodic?
          periodic_events, new_periodic_last_epoch = get_periodic_events
          events = periodic_events || []
        end
        if singular?
          singular_events, new_singular_last_epoch = get_singular_events
          events.nil? ? (events = singular_events) : events.concat(singular_events || [])
        end

        $logger.debug {
          msgs = ["Finish process #{resource.uri}"]
          msgs << "periodic_last_epoch:#{periodic_last_epoch}" if periodic_last_epoch
          msgs << "singular_last_epoch:#{singular_last_epoch}" if singular_last_epoch
          msgs << "new_periodic_last_epoch:#{new_periodic_last_epoch}" if new_periodic_last_epoch
          msgs << "new_singular_last_epoch:#{new_singular_last_epoch}" if new_singular_last_epoch
          msgs.join(', ')
        }

        return nil if events.nil? || events.empty?
        yield(events) if block_given? # send_message
        update_status_file(:periodic_last_epoch, new_periodic_last_epoch) if new_periodic_last_epoch
        update_status_file(:singular_last_epoch, new_singular_last_epoch) if new_singular_last_epoch
        true
      end

      def get_periodic_events
        if hourly?
          events, new_last_epoch, rows = get_hourly_events
          if daily?
            daily_events = build_daily_events_from_hourly(rows)
            events.concat(daily_events)
          end
          [events, new_last_epoch]
        elsif daily?
          get_daily_events
        else
          raise
        end
      end

      def get_singular_events
        sql = "select " \
          "NULL AS d, NULL AS h, max(epoch) " \
          "from #{q_db}.#{q_schema}.#{q_table} " \
          "#{q_where.empty? ? '' : "where #{q_where} "}" \
          "having max(epoch) > #{q_singular_last_epoch}"
        query_and_get_events(:singular, sql)
      end

      def get_hourly_events
        sql = "select " \
          "#{q_date} AS d, DATE_PART('hour', #{q_timestamp}) AS h, max(epoch) " \
          "from #{q_db}.#{q_schema}.#{q_table} " \
          "where #{q_date} IN ('#{dates.join("','")}') " \
          "#{q_where.empty? ? '' : "AND #{q_where} "}" \
          "group by d, h having max(epoch) > #{q_periodic_last_epoch} " \
          "order by d, h"
        query_and_get_events(:hourly, sql)
      end

      def get_daily_events
        sql = "select " \
          "#{q_date} AS d, 0 AS h, max(epoch) " \
          "from #{q_db}.#{q_schema}.#{q_table} " \
          "where #{q_date} IN ('#{dates.join("','")}') " \
          "#{q_where.empty? ? '' : "AND #{q_where} "}" \
          "group by d having max(epoch) > #{q_periodic_last_epoch} " \
          "order by d"
        query_and_get_events(:daily, sql)
      end

      private

      def query_and_get_events(unit, sql)
        $logger.debug { "Query: #{sql}" }
        rows = connection.query(sql)
        events = build_events(unit, rows)
        new_last_epoch = build_latest_epoch(rows)
        [events, new_last_epoch, rows]
      rescue ::Vertica::Error::QueryError => e
        $logger.warn { "#{e.class} #{e.message}" } # e.message includes sql
        nil
      rescue ::Vertica::Error::TimedOutError => e
        $logger.warn { "#{e.class} #{e.message} SQL:#{sql}" }
        nil
      end

      def update_status_file(key, last_epoch)
        @status.set(key.to_sym, last_epoch)
      end

      def get_from_status_file(key)
        @status.getsetnx(key.to_sym, get_current_epoch)
      end

      def get_current_epoch
        connection.query('select GET_CURRENT_EPOCH()').first.first
      end

      def resource_valid?
        resource_unit_valid? && !resource.timezone.nil? && !resource.span_in_days.nil?
      end

      def resource_unit_valid?
        resource.unit.split(',').each do |item|
          return false unless %w[singular daily hourly].include?(item)
        end
        true
      end

      def hourly?
        return @is_hourly unless @is_hourly.nil?
        @is_hourly = resource.unit.include?('hourly')
      end

      def daily?
        return @is_daily unless @is_daily.nil?
        @is_daily = resource.unit.include?('daily')
      end

      def singular?
        return @is_singular unless @is_singular.nil?
        @is_singular = resource.unit.include?('singular')
      end

      def periodic?
        hourly? or daily?
      end

      def dates
        return @dates if @dates
        now = Time.now.localtime(resource.timezone)
        @dates = resource.span_in_days.times.map do |i|
          (now - (i * 86400)).strftime('%Y-%m-%d')
        end
      end

      def build_latest_epoch(rows)
        rows.map {|row| row[2] }.max
      end

      def build_events(unit, rows)
        rows.map do |row|
          date, hour, epoch = row[0], row[1], row[2]
          {
            uuid: SecureRandom.uuid,
            resource_uri: resource.uri,
            resource_unit: unit.to_s,
            resource_time: date_hour_to_i(date, hour, resource.timezone),
            resource_timezone: resource.timezone,
            payload: (date ? {d: date.to_s, h: hour.to_i} : {}).merge!(epoch: epoch).to_json,
          }
        end
      end

      def build_daily_events_from_hourly(rows)
        max_epoch_of = {}
        rows.each do |row|
          date, hour, epoch = row[0], row[1], row[2]
          max_epoch_of[date] = [epoch, max_epoch_of[date] || 0].max
        end
        daily_events = max_epoch_of.map do |date, epoch|
          {
            uuid: SecureRandom.uuid,
            resource_uri: resource.uri,
            resource_unit: 'daily',
            resource_time: date_hour_to_i(date, 0, resource.timezone),
            resource_timezone: resource.timezone,
            payload: {d: date.to_s, h: 0, epoch: epoch}.to_json,
          }
        end
      end

      def date_hour_to_i(date, hour, timezone)
        return 0 if date.nil?
        Time.strptime("#{date.strftime("%Y-%m-%d")} #{hour.to_i} #{timezone}", '%Y-%m-%d %H %z').to_i
      end

      def q_periodic_last_epoch
        @q_periodic_last_epoch ||= ::Vertica.quote(periodic_last_epoch)
      end

      def q_singular_last_epoch
        @q_singular_last_epoch ||= ::Vertica.quote(singular_last_epoch)
      end

      def parsed_uri
        @parsed_uri ||= URI.parse(resource.uri)
      end

      def parsed_query
        @parsed_query ||= Rack::Utils.parse_nested_query(parsed_uri.query)
      end

      def db
        @db ||= parsed_uri.path[1..-1].split('/')[0]
      end

      def schema
        @schema ||= parsed_uri.path[1..-1].split('/')[1]
      end

      def table
        @table ||= parsed_uri.path[1..-1].split('/')[2]
      end

      def date_column
        parsed_query['date'] || $setting.dig(:vertica, :date_column) || 'd'
      end

      def timestamp_column
        parsed_query['timestamp'] || $setting.dig(:vertica, :timestamp_column) || 't'
      end

      def where
        parsed_query['where'] || {}
      end

      def q_db
        @q_db ||= ::Vertica.quote_identifier(db)
      end

      def q_schema
        @q_schema ||= ::Vertica.quote_identifier(schema)
      end

      def q_table
        @q_table ||= ::Vertica.quote_identifier(table)
      end

      def q_date
        @q_date ||= ::Vertica.quote_identifier(date_column)
      end

      def q_timestamp
        @q_timestamp ||= ::Vertica.quote_identifier(timestamp_column)
      end

      # Value specification:
      # * A value looks like an integer string is treated as an integer.
      # * If you want to treat it as as string, surround with double quote or single quote.
      # * A value does not look like an integer is treated as a string.
      # Operator specification:
      # * Only equality is supported now
      def q_where
        @q_where ||= where.map do |col, val|
          begin
            val = Integer(val)
          rescue => e
            if val.start_with?("'") and val.end_with?("'")
              val = val[1..-2]
            elsif val.start_with?('"') and val.end_with?('"')
              val = val[1..-2]
            end
          end
          "#{::Vertica.quote_identifier(col)} = #{::Vertica.quote(val)}"
        end.join(' AND ')
      end
    end
  end
end
